### PR TITLE
Read object features table

### PR DESCRIPTION
**Pull request by emilmelnikov**
Add the ability to read an object features table from object classification workflow.

Also, refactor and simplify all workflows and associated demos.

Closes https://github.com/ilastik/ilastik4ij/issues/56.

### CODE REVIEW

**Review by wolny — APPROVED** [review]
This is very nicely done, great job @emilmelnikov! The code is much cleaner than what it was before and I'm glad that the workflows can finally be implemented using the `@Plugin` annotations, instead of the oldschool `GenericCommand` (?). 
:+1: :shipit:

**Comment by k-dominik** [comment]
Are these changes backwards compatible (e.g. will old macros still run)? I've tried some of my old ones, but so far without success.

File: src/main/java/org/ilastik/ilastik4ij/ui/OptionsIlastik.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import org.scijava.options.OptionsPlugin;
+import org.scijava.plugin.Parameter;
+import org.scijava.plugin.Plugin;
+import org.scijava.widget.FileWidget;
+
+import java.io.File;
+
+@Plugin(type = OptionsPlugin.class, menuPath = "Plugins>ilastik>Configure ilastik executable location")
+public class OptionsIlastik extends OptionsPlugin {
+    @Parameter(label = "ilastik executable path", style = FileWidget.OPEN_STYLE)
```
Comment (k-dominik) [inline src/main/java/org/ilastik/ilastik4ij/ui/OptionsIlastik.java:13]: was previously named `path to ilastik executable`. The rename results in not picking up the "old" value of that variable

Comment (k-dominik) [reply src/main/java/org/ilastik/ilastik4ij/ui/OptionsIlastik.java:13]: well, not sure if it is _this_ rename. in any case, usually it picks up the executable after the update, not this time

File: src/main/java/org/ilastik/ilastik4ij/workflow/WorkflowCommand.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    public LogService logService;
+
+    @Parameter
+    public StatusService statusService;
+
+    @Parameter
+    public OptionsService optionsService;
+
+    @Parameter(label = "Trained ilastik project file")
+    public File project;
```
Comment (k-dominik) [inline src/main/java/org/ilastik/ilastik4ij/workflow/WorkflowCommand.java:38]:
```suggestion
    public File projectfilename;
```

File: src/main/java/org/ilastik/ilastik4ij/workflow/WorkflowCommand.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    @Parameter(label = "Raw input image")
+    public Dataset rawData;
```
Comment (k-dominik) [inline src/main/java/org/ilastik/ilastik4ij/workflow/WorkflowCommand.java:41]:
```suggestion
    public Dataset inputimage;
```

File: src/main/java/org/ilastik/ilastik4ij/workflow/ObjectClassificationCommand.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+
+@Plugin(type = Command.class, headless = true, menuPath = "Plugins>ilastik>Run Object Classification Prediction")
+public final class ObjectClassificationCommand<T extends NativeType<T>> extends WorkflowCommand<T> {
+    @Parameter(label = "Pixel probability or segmentation image")
+    public Dataset input2;
```
Comment (k-dominik) [inline src/main/java/org/ilastik/ilastik4ij/workflow/ObjectClassificationCommand.java:21]:
```suggestion
    public Dataset inputProbOrSegImage;
```

**Comment by emilmelnikov** [comment]
> Are these changes backwards compatible (e.g. will old macros still run)? I've tried some of my old ones, but so far without success.
> 
> for Pixel Classification it seems that parameters have been renamed:
> 
> * `projectfilename` -> `project`
> * `inputimage` -> `rawdata`
> 
> I know this sucks, I like the more expressive parameter names, too, but we can't break old macros because of an internal refactor

Good call, I forgot about macros.

File: src/main/java/org/ilastik/ilastik4ij/ui/OptionsIlastik.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Parameter(min = "-1", label = "maximum number of threads (\"-1\" means no restriction)")
+    private int numThreads = -1;
+
+    @Parameter(min = "256", label = "maximum amount of megabytes")
+    private int maxRamMb = 4096;
+
+    @SuppressWarnings("unused")
+    @Parameter(visibility = ItemVisibility.MESSAGE)
+    private final String examples = "<html>ilastik executable path examples:<dl>" +
```
Comment (k-dominik) [inline src/main/java/org/ilastik/ilastik4ij/ui/OptionsIlastik.java:24]: We did have this "feature" that only the string relevant for the OS would be shown, any reason you removed that?

**Comment by k-dominik** [comment]
Just to add now that I looked at the code it is really nice to see reduced amount of duplication and boilerplate for the workflows! I tested generating the tables and this works nicely as well. We'll have to add documentation for this though, as one has to configure the csv export in ilastik first - otherwise it won't work. The real power of the table can be unleashed once we allow for exporting the _Object Identities_. That way (I think) one could use feature table as colormaps (somehow) :D. We should maybe ask @tischi how to best do it and provide an example.

**Comment by emilmelnikov** [comment]
> We'll have to add documentation for this though, as one has to configure the csv export in ilastik first - otherwise it won't work.

Would this work?

Screenshot 2021-05-31 at 11 57 45

File: src/main/java/org/ilastik/ilastik4ij/ui/OptionsIlastik.java (hunk 1 above)
Comment (emilmelnikov) [reply src/main/java/org/ilastik/ilastik4ij/ui/OptionsIlastik.java]: Rolled back.

File: src/main/java/org/ilastik/ilastik4ij/ui/OptionsIlastik.java (hunk 5 above)
Comment (emilmelnikov) [reply src/main/java/org/ilastik/ilastik4ij/ui/OptionsIlastik.java]: Wanted to use less magic, but fair point. Now the the correct help string is generated in a static initializer.

**Comment by emilmelnikov** [comment]
> Just to add now that I looked at the code it is really nice to see reduced amount of duplication and boilerplate for the workflows! I tested generating the tables and this works nicely as well. We'll have to add documentation for this though, as one has to configure the csv export in ilastik first - otherwise it won't work. The real power of the table can be unleashed once we allow for exporting the _Object Identities_. That way (I think) one could use feature table as colormaps (somehow) :D. We should maybe ask @tischi how to best do it and provide an example.

I presume we need to change table export in ilastik for this?

**Comment by k-dominik** [comment]
> > The real power of the table can be unleashed once we allow for exporting the _Object Identities_.
> 
> I presume we need to change table export in ilastik for this?

no, we'd need to allow exporting a different source via `--export_source=Object Identities"` command line argument.



> > We'll have to add documentation for this though, as one has to configure the csv export in ilastik first - otherwise it won't work.
> 
> Would this work?
> 
> image

that's perfect :) users don't even need to look into some docs.

**Comment by emilmelnikov** [comment]
> we'd need to allow exporting a different source via `--export_source=Object Identities"` command line argument.

Should we address this in a separate PR?

**Comment by emilmelnikov** [comment]
Superseded by https://github.com/ilastik/ilastik4ij/pull/106.

Object features table could be addressed in a separate PR.

**Pull request closed without merging**